### PR TITLE
Fix a bug with diophantine() when cornacchia() is not called properly

### LINKED ISSUE #18628: diophantine(), diop_quadratic() missing solutions
With sympy version 1.5.1:

```
from sympy import *
from sympy.solvers.diophantine import diop_quadratic

a, b, x, y = symbols('a b x y', integer=True, positive=True)
f = -a*x + x**2 - b*y + y**2
g = f.subs([(a, 15), (b, 8)])
print('diophantine   (', g, ') solutions:', diophantine(g))
print('diop_quadratic(', g, ') solutions:', diop_quadratic(g))
h = f.subs([(a, 8), (b, 15)])
print('diophantine   (', h, ') solutions:', diophantine(h))
print('diop_quadratic(', h, ') solutions:', diop_quadratic(h))
print(g.subs([(x, 16), (y, 4)]))

```
Output:

```
diophantine   ( x**2 - 15*x + y**2 - 8*y ) solutions: {(15, 8)}
diop_quadratic( x**2 - 15*x + y**2 - 8*y ) solutions: {(15, 0), (0, 8), (15, 8), (0, 0)}
diophantine   ( x**2 - 8*x + y**2 - 15*y ) solutions: {(8, 15), (4, 16)}
diop_quadratic( x**2 - 8*x + y**2 - 15*y ) solutions: {(0, 0), (0, 15), (8, 0), (4, 16), (4, -1), (8, 15)}
0

```
First, I expect diophantine() to return the same solutions as diop_quadratic(), but it doesn't.

Second, the function f() is symmetric in x and y, so the solutions to g() and h() should be symmetric, but they aren't.  For example, (16, 4) is a solution to g(), but is missing from the solution set.

### PR DESCRIPTION
**Pull request by zylipku**
Occasionally diophantine() and diop_quadratic() result in a DN form $X^2 - DY^2 = N$ which cannot be handled by the current cornacchia() properly. For example, previously the diophantine solver will miss some solutions:
```
>>> eq1 = x**2 - 15*x + y**2 - 8*y # also mentioned in #18628
>>> diophantine(eq1)
{(0, 0), (15, 0), (15, 8), (0, 8)}
>>> eq1.subs({x: 16, y: 4})
0
```
```
>>> eq2 = 2*x**2 - 9*x + 4*y**2 - 8*y + 14
>>> diophantine(eq2)
set()
>>> eq2.subs({x: 2, y: 1})
0
```
Both the two examples lead to a function call of cornacchia(1, X, 1) with X > 1, which incorrrectly turns out to have no solution by using cornacchia(). The solver can only be applied when $a+b\le m$. See reference from docs:
http://www.numbertheory.org/php/cornacchia.html 
An additional check has been added before the algorithm based on the fact that $ax^2 + by^2 = m$ admits a solution for $a + b > m$ only if $x = 0$ or $y = 0$. A test function including both the two examples has been appended as well.

#### References to other Issues or PRs

Fixes #18628 

#### Brief description of what is fixed or changed
An additional check has been added before the cornacchia algorithm body based on the fact that $ax^2 + by^2 = m$ admits a solution for $a + b > m$ only if $x = 0$ or $y = 0$.

#### Release Notes

* solvers
  * Fix a bug with cornacchia(). Formerly diophantine() and diop_quadratic() may miss some solutions for quadratic equations.

### CODE REVIEW

File: sympy/solvers/diophantine/diophantine.py [hunk 1]
```diff
@@ -2174,6 +2174,25 @@ def cornacchia(a:int, b:int, m:int) -> set[tuple[int, int]]:
     """
     # Assume gcd(a, b) = gcd(a, m) = 1 and a, b > 0 but no error checking
     sols = set()
+
+    if a + b > m:
+        # xy = 0 must hold if there exists a solution
+        if m % a == 0:
```
Comment (haru-44) [inline sympy/solvers/diophantine/diophantine.py:2180]: Given the constraint `gcd(a, m) = 1`, I think the condition `if a == 1` should be sufficient.

File: sympy/solvers/diophantine/diophantine.py [hunk 2]
```diff
@@ -2174,6 +2174,25 @@ def cornacchia(a:int, b:int, m:int) -> set[tuple[int, int]]:
...
+            # y = 0
+            s, _exact = iroot(m // a, 2)
+            if _exact:
+                sols.add((int(s), int(0)))
```
Comment (haru-44) [inline sympy/solvers/diophantine/diophantine.py:2184]:
```suggestion
                sols.add((int(s), 0))
```

File: sympy/solvers/diophantine/diophantine.py [hunk 3]
```diff
@@ -2174,6 +2174,25 @@ def cornacchia(a:int, b:int, m:int) -> set[tuple[int, int]]:
...
+            if a == b:
+            # only keep one solution
```
Comment (haru-44) [inline sympy/solvers/diophantine/diophantine.py:2186]:
```suggestion
                # only keep one solution
```

File: sympy/solvers/diophantine/diophantine.py [hunk 4]
```diff
@@ -2174,6 +2174,25 @@ def cornacchia(a:int, b:int, m:int) -> set[tuple[int, int]]:
...
+                return sols
+        if m % b == 0:
+            # x = 0
+            s, _exact = iroot(m // b, 2)
+            if _exact:
+                sols.add((int(0), int(s)))
```
Comment (haru-44) [inline sympy/solvers/diophantine/diophantine.py:2192]:
```suggestion
                sols.add((0, int(s)))
```

**Comment by haru-44** [comment]
thanks! looks good.

**Comment by oscarbenjamin** [comment]
Thanks

**Pull request merged**